### ISSUE TITLE
CA0.6.1 AWS - able to scale down to 0 but back up

### ISSUE THREAD

**Issue by innovia** [kubernetes/autoscaler#346]
Hi

Im using kops on AWS with CA 0.6.1

I am able to scale down to zero but not back up from zero

also not the format of "No pod can fit to %sgpu-nodes.server.com" the %s is printed - i think its only on the logs...

```
I0914 20:59:01.756751       1 static_autoscaler.go:213] Filtering out schedulables
I0914 20:59:01.756852       1 static_autoscaler.go:221] No schedulable pods
I0914 20:59:01.756869       1 scale_up.go:50] Pod default/ami-skysegment-l1knx is unschedulable
W0914 20:59:01.855358       1 aws_manager.go:201] Found multiple availability zones, using us-east-1a
I0914 20:59:01.855580       1 scale_up.go:71] Upcoming 0 nodes
I0914 20:59:01.904324       1 scale_up.go:112] Scale-up predicate failed: GeneralPredicates predicate mismatch, cannot put default/ami-skysegment-l1knx on template-node-for-gpu-nodes.server.com-6748296357877885074, reason: Insufficient alpha.kubernetes.io/nvidia-gpu
I0914 20:59:01.904350       1 scale_up.go:141] No pod can fit to %sgpu-nodes.server.com
I0914 20:59:01.904359       1 scale_up.go:146] No expansion options
I0914 20:59:01.904396       1 static_autoscaler.go:270] Scale down status: unneededOnly=false lastScaleUpTime=2017-09-07 07:13:31.488444337 +0000 UTC lastScaleDownFailedTrail=2017-09-03 10:53:59.066502018 +0000 UTC schedulablePodsPresent=false
I0914 20:59:01.904412       1 static_autoscaler.go:272] Calculating unneeded nodes
I0914 20:59:01.904463       1 utils.go:360] Skipping ip-172-10-100-72.ec2.internal - no node group config
I0914 20:59:01.904496       1 utils.go:360] Skipping ip-172-10-165-18.ec2.internal - no node group config
I0914 20:59:01.904572       1 static_autoscaler.go:292] Starting scale down
I0914 20:59:01.904588       1 scale_down.go:260] No candidates for scale down
W0914 20:59:01.904600       1 clusterstate.go:271] Failed to find readiness information for gpu-nodes.server.com
I0914 20:59:01.905081       1 event.go:218] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"ami-skysegment-l1knx", UID:"6db6c815-998f-11e7-ac58-06fcf22dd9b0", APIVersion:"v1", ResourceVersion:"20001417", FieldPath:""}): type: 'Normal' reason: 'NotTriggerScaleUp' pod didn't trigger scale-up (it wouldn't fit if a new node is added)
```

my command on CA pod is
```
./cluster-autoscaler
--v=4
--stderrthreshold=info
--cloud-provider=aws
--skip-nodes-with-local-storage=false
--nodes=0:10:gpu-nodes.server.com
```
my node config

```yaml

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2017-05-15T11:01:53Z
  labels:
    kops.k8s.io/cluster: server.com
  name: gpu-nodes
spec:
  cloudLabels:
    gpu_ml: spot
    k8s.io/cluster-autoscaler/node-template/label/node_type: gpu
    service_name: machine-learning
  image: k8s-1.6-alpha-debian-jessie-amd64-hvm-ebs-2017-05-15
  machineType: p2.xlarge
  maxPrice: "0.40"
  maxSize: 10
  minSize: 0
  role: Node
  rootVolumeSize: 30
  rootVolumeType: gp2
  subnets:
  - us-east-1a
  - us-east-1b
  - us-east-1c
  - us-east-1d
  - us-east-1e
  - us-east-1f
```

Would appreciate help in resolving this as this will help us a lot.
Thanks

Ami

**Comment by mwielgus** [comment]
Can you share the pending pod definition (`kubectl get pod <name> -o yaml`) and a sample node (`kubectl get node <name> -o yaml`) from ASG that you expected to grow from 0.

**Comment by innovia** [comment]
@mwielgus 

I'm not sure if editing the auto scale group either via kops or via AWS Console confuse the cluster autosacaler.

The issue is that I am able to scale down to zero and scale up from 1 node but not from zero

here's what happens when I attempted to launch another pod (insufficient resource) when there's already 1 node up to begin with.

also after deleting the 2 jobs and pods the CA says it's starting to scale down unneeded nodes but nothing happens even if I manually set the minimum to 0 (2 desired: 0 min: 10 max)

The only thing Im thinking of is the setting for CA to find the ASG must match the actual settings of the ASG *before* the CA run?

```
I0915 18:57:17.460280       1 scale_up.go:251] Scale-up: setting group gpu-nodes.server.com size to 2
I0915 18:57:17.518402       1 aws_manager.go:126] Setting asg gpu-nodes.server.com size to 2
I0915 18:57:17.559739       1 event.go:218] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"ami-skysegment-2nd-test-mr0k2", UID:"af552921-9a47-11e7-ac58-06fcf22dd9b0", APIVersion:"v1", ResourceVersion:"20142817", FieldPath:""}): type: 'Normal' reason: 'TriggeredScaleUp' pod triggered scale-up: [{gpu-nodes.server.com 1->2 (max: 10)}]
I0915 18:57:17.559762       1 event.go:218] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"5530c58c-907c-11e7-8f1a-063f7f4455a2", APIVersion:"v1", ResourceVersion:"20142802", FieldPath:""}): type: 'Normal' reason: 'ScaledUpGroup' Scale-up: group gpu-nodes.server.com size set to 2
I0915 18:57:18.982639       1 leaderelection.go:204] successfully renewed lease kube-system/cluster-autoscaler
I0915 18:57:20.989719       1 leaderelection.go:204] successfully renewed lease kube-system/cluster-autoscaler
I0915 18:57:22.997314       1 leaderelection.go:204] successfully renewed lease kube-system/cluster-autoscaler
I0915 18:57:25.067187       1 leaderelection.go:204] successfully renewed lease kube-system/cluster-autoscaler
I0915 18:57:27.160031       1 leaderelection.go:204] successfully renewed lease kube-system/cluster-autoscaler
I0915 18:57:27.689779       1 static_autoscaler.go:213] Filtering out schedulables
I0915 18:57:27.690377       1 static_autoscaler.go:221] No schedulable pods
I0915 18:57:27.690433       1 scale_up.go:50] Pod default/ami-skysegment-2nd-test-mr0k2 is unschedulable
I0915 18:57:27.698524       1 scale_up.go:71] Upcoming 1 nodes
I0915 18:57:27.726748       1 scale_up.go:138] No need for any nodes in gpu-nodes.server.com
```

here's the job yaml config:
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: ami-skysegment
spec:
  activeDeadlineSeconds: 1200
  completions: 1
  parallelism: 1
  template:
    metadata:
      name: ami-skysegment
      labels:
        user_name: ami
        commit_sha1: 1af15217c265a2c41d55d491a10d9abc5a05a848
        output_folder: ami-14-09-2017T22-57-47-da0ca459372a-skysegment
        uuid: da0ca459372a
    spec:
      restartPolicy: OnFailure
      initContainers:
      - name: init
        image: "tensorflow/tensorflow"
        command:
        - "/bin/bash"
        - "-c"
        - |-
          exec &> >(tee -a /code/init_container.log)
          set -e
          set -o pipefail
          echo "Bootstrap started"
          echo "Setting up code base"
          ...
          echo "Bootstrap completed."
        volumeMounts:
        - name: output
          mountPath: /output
        - name: code
          mountPath: /code
      containers:
      - image: "tensorflow/tensorflow"
        resources:
          limits:
            alpha.kubernetes.io/nvidia-gpu: 1
        workingDir: "/code"
        name: tensorflow-worker
        command:
        - "/bin/bash"
        - "-c"
        - |-
          exec &> >(tee -a /output/ami/14_09_2017T22_57_47_da0ca459372a_skysegment/job.log)
          set -e
          set -o pipefail
          echo "Copying Nvidia and Cuda Lib support into the container"
          cp -d /rootfs/usr/lib/x86_64-linux-gnu/libcuda.* /usr/lib/x86_64-linux-gnu/
          cp -d /rootfs/usr/lib/x86_64-linux-gnu/libnvidia* /usr/lib/x86_64-linux-gnu/
          ...
          echo "Job done"
        volumeMounts:
        - name: rootfs-usr-lib
          mountPath: /rootfs/usr/lib
        - name: input
          mountPath: /input
          readOnly: true
        - name: output
          mountPath: /output
        - name: code
          mountPath: /code
        - name: send-email-configmap
          mountPath: /etc/ses
        env:
        - name: USER_NAME
          value: ami
        - name: JOBNAME
          value: skysegment
        - name: OUTPUT
          value: /output/ami/14_09_2017T22_57_47_da0ca459372a_skysegment
        - name: COMMIT
          value: 1af15217c265a2c41d55d491a10d9abc5a05a848
        - name: VERBOSE
          value: "false"

      volumes:
      - name: rootfs-usr-lib
        hostPath:
          path: /usr/lib
      - name: input
        persistentVolumeClaim:
          claimName: ml-input-nfs-pvc
      - name: output
        persistentVolumeClaim:
          claimName: ml-output-nfs-pvc
      - name: code
        emptyDir: {}
      - name: send-email-configmap
        configMap:
          name: ses-send-email-script-configmap
```

and here is a GPU node config yaml 

```yaml
apiVersion: v1
kind: Node
metadata:
  annotations:
    node.alpha.kubernetes.io/ttl: "0"
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: 2017-09-15T18:50:56Z
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: p2.xlarge
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east-1
    failure-domain.beta.kubernetes.io/zone: us-east-1e
    kubernetes.io/hostname: ip-172-20-182-148.ec2.internal
    kubernetes.io/role: node
    node-role.kubernetes.io/node: ""
  name: ip-172-20-182-148.ec2.internal
  resourceVersion: "20142380"
  selfLink: /api/v1/nodes/ip-172-20-182-148.ec2.internal
  uid: ced7c83c-9a46-11e7-ac58-06fcf22dd9b0
spec:
  externalID: i-052a9bd94e698f37e
  podCIDR: 100.96.14.0/24
  providerID: aws:///us-east-1e/i-052a9bd94e698f37e
status:
  addresses:
  - address: 172.20.182.148
    type: InternalIP
  - address: 52.87.233.13
    type: ExternalIP
  - address: ip-172-20-182-148.ec2.internal
    type: InternalDNS
  - address: ec2-52-87-233-13.compute-1.amazonaws.com
    type: ExternalDNS
  - address: ip-172-20-182-148.ec2.internal
    type: Hostname
  allocatable:
    alpha.kubernetes.io/nvidia-gpu: "1"
    cpu: "4"
    memory: 62781872Ki
    pods: "110"
  capacity:
    alpha.kubernetes.io/nvidia-gpu: "1"
    cpu: "4"
    memory: 62884272Ki
    pods: "110"
  conditions:
  - lastHeartbeatTime: 2017-09-15T18:51:07Z
    lastTransitionTime: 2017-09-15T18:51:07Z
    message: RouteController created a route
    reason: RouteCreated
    status: "False"
    type: NetworkUnavailable
  - lastHeartbeatTime: 2017-09-15T18:53:26Z
    lastTransitionTime: 2017-09-15T18:50:56Z
    message: kubelet has sufficient disk space available
    reason: KubeletHasSufficientDisk
    status: "False"
    type: OutOfDisk
  - lastHeartbeatTime: 2017-09-15T18:53:26Z
    lastTransitionTime: 2017-09-15T18:50:56Z
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: 2017-09-15T18:53:26Z
    lastTransitionTime: 2017-09-15T18:50:56Z
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: 2017-09-15T18:53:26Z
    lastTransitionTime: 2017-09-15T18:51:16Z
    message: kubelet is posting ready status
    reason: KubeletReady
    status: "True"
    type: Ready
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  images:
  - names:
    - 123456789012.dkr.ecr.us-west-2.amazonaws.com/fluentd-elasticsearch-aws@sha256:48f165216f40340d1253947b32cdf026a908876c1f836aa595bd5346216c7855
    - 123456789012.dkr.ecr.us-west-2.amazonaws.com/fluentd-elasticsearch-aws:1.0.1
    sizeBytes: 462145895
  - names:
    - protokube:1.7.0
    sizeBytes: 363365587
  - names:
    - gcr.io/google_containers/kube-proxy@sha256:2c6f2b7b3d1a93c1e238a5ee4d0a2d9f7bcdad607254b99d4489ec920d26ca01
    - gcr.io/google_containers/kube-proxy:v1.7.0
    sizeBytes: 114728022
  - names:
    - gcr.io/google_containers/pause-amd64@sha256:163ac025575b775d1c0f9bf0bdd0f086883171eb475b5068e7defa4ca9e76516
    - gcr.io/google_containers/pause-amd64:3.0
    sizeBytes: 746888
  nodeInfo:
    architecture: amd64
    bootID: 4dd1c9d4-94be-495f-b3d8-6358d942e75d
    containerRuntimeVersion: docker://1.12.6
    kernelVersion: 4.4.41-k8s
    kubeProxyVersion: v1.7.0
    kubeletVersion: v1.7.0
    machineID: 7768131d5eea4beaa5f3950f16d7f47e
    operatingSystem: linux
    osImage: Debian GNU/Linux 8 (jessie)
    systemUUID: EC267F8E-B99D-0C51-2238-C7CA4F8423A6
```

and here's the regular node describe
```
Name:			ip-172-20-182-148.ec2.internal
Role:			
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=p2.xlarge
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=us-east-1
			failure-domain.beta.kubernetes.io/zone=us-east-1e
			kubernetes.io/hostname=ip-172-20-182-148.ec2.internal
			kubernetes.io/role=node
			node-role.kubernetes.io/node=
Annotations:		node.alpha.kubernetes.io/ttl=0
			volumes.kubernetes.io/controller-managed-attach-detach=true
Taints:			<none>
CreationTimestamp:	Fri, 15 Sep 2017 21:50:56 +0300
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  NetworkUnavailable 	False 	Fri, 15 Sep 2017 21:51:07 +0300 	Fri, 15 Sep 2017 21:51:07 +0300 	RouteCreated 			RouteController created a route
  OutOfDisk 		False 	Fri, 15 Sep 2017 21:52:36 +0300 	Fri, 15 Sep 2017 21:50:56 +0300 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Fri, 15 Sep 2017 21:52:36 +0300 	Fri, 15 Sep 2017 21:50:56 +0300 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Fri, 15 Sep 2017 21:52:36 +0300 	Fri, 15 Sep 2017 21:50:56 +0300 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		True 	Fri, 15 Sep 2017 21:52:36 +0300 	Fri, 15 Sep 2017 21:51:16 +0300 	KubeletReady 			kubelet is posting ready status
Addresses:
  InternalIP:	172.20.182.148
  ExternalIP:	52.87.233.13
  InternalDNS:	ip-172-20-182-148.ec2.internal
  ExternalDNS:	ec2-52-87-233-13.compute-1.amazonaws.com
  Hostname:	ip-172-20-182-148.ec2.internal
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	1
 cpu:					4
 memory:				62884272Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	1
 cpu:					4
 memory:				62781872Ki
 pods:					110
System Info:
 Machine ID:			7768131d5eea4beaa5f3950f16d7f47e
 System UUID:			EC267F8E-B99D-0C51-2238-C7CA4F8423A6
 Boot ID:			4dd1c9d4-94be-495f-b3d8-6358d942e75d
 Kernel Version:		4.4.41-k8s
 OS Image:			Debian GNU/Linux 8 (jessie)
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.12.6
 Kubelet Version:		v1.7.0
 Kube-Proxy Version:		v1.7.0
PodCIDR:			100.96.14.0/24
ExternalID:			i-052a9bd94e698f37e
Non-terminated Pods:		(2 in total)
  Namespace			Name							CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----							------------	----------	---------------	-------------
  kube-system			fluentd-elasticsearch-pph61				100m (2%)	0 (0%)		200Mi (0%)	200Mi (0%)
  kube-system			kube-proxy-ip-172-20-182-148.ec2.internal		100m (2%)	0 (0%)		0 (0%)		0 (0%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  200m (5%)	0 (0%)		200Mi (0%)	200Mi (0%)
Events:
  FirstSeen	LastSeen	Count	From						SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----						-------------	--------	------			-------
  1m		1m		1	kubelet, ip-172-20-182-148.ec2.internal				Normal		Starting		Starting kubelet.
  1m		1m		2	kubelet, ip-172-20-182-148.ec2.internal				Normal		NodeHasSufficientDisk	Node ip-172-20-182-148.ec2.internal status is now: NodeHasSufficientDisk
  1m		1m		2	kubelet, ip-172-20-182-148.ec2.internal				Normal		NodeHasSufficientMemory	Node ip-172-20-182-148.ec2.internal status is now: NodeHasSufficientMemory
  1m		1m		2	kubelet, ip-172-20-182-148.ec2.internal				Normal		NodeHasNoDiskPressure	Node ip-172-20-182-148.ec2.internal status is now: NodeHasNoDiskPressure
  1m		1m		1	kubelet, ip-172-20-182-148.ec2.internal				Normal		NodeAllocatableEnforced	Updated Node Allocatable limit across pods
  1m		1m		1	kube-proxy, ip-172-20-182-148.ec2.internal			Normal		Starting		Starting kube-proxy.
  1m		1m		1	kubelet, ip-172-20-182-148.ec2.internal				Normal		NodeReady		Node ip-172-20-182-148.ec2.internal status is now: NodeReady

```

**Comment by mwielgus** [comment]
Thanks for the dumps. It wasn't obvious enough that your pods required GPUs.

CA in version 0.6.1 does is not aware of GPUs in AWS. However every time CA sees a pending pod it needs to create an internal "template" node to tell whether the scale-up make sense or not. It is simpler when there is at least one node present for each of the ASG. If there is no, then CA has to guess what the node would look like. And prior to fix https://github.com/kubernetes/autoscaler/pull/322 it didn't put the correct information. 

We can cherry-pick #322 for you to 0.6 branch and it will be released with 0.6.3 in ~2 week. Of course you can build yourself the fixed version earlier. 
You can also wait for 0.7 which will be released in ~2 week, which will contain the GPU fix as well as couple other fixes and improvements.

**Comment by innovia** [comment]
Thanks ill try to build it myslef

So let me try and understand exactly how it works

CA is fed with the name of the ASG but it has to understand how a pod is belong to that group?

**Comment by innovia** [comment]
@mwielgus can you also provide instructions no how to build?

**Comment by mwielgus** [comment]
Please check https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/Makefile.

You need to specify `TAG` and `REGISTRY` env variables. 
Modify  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/Makefile#L30.
to match your registry.
Then `make release`.

**Comment by mwielgus** [comment]
As of why we need node templates please take a look at:

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-up-work

**Comment by innovia** [comment]
Looks like p2 instances are configured in that PR as 0 gpu...

**Comment by mwielgus** [comment]
@innovia Please check https://github.com/kubernetes/autoscaler/issues/321#issuecomment-329007142

**Comment by innovia** [comment]
thanks for all the help!

Built using  master and made a manual fix for the p2.xlarge 

gen.go line 147
```go
if attr.InstanceType == "p2.xlarge" {
						instanceTypes[attr.InstanceType].GPU = 1
					}
```

then running go generate ./cloudprovider/aws followed by make release with env var for REPOSITORY and TAG

**Comment by 7chenko** [comment]
@innovia This fix didn't work for me, still getting "pod didn't trigger scale-up (it wouldn't fit if a new node is added)" for p2.xlarge instances, not sure why.

I do have a support case open with AWS for them to fix their json file.

**Comment by innovia** [comment]
Make sure you bump the version of both the cluster autoscaler as well as the docker you build

Try runing make clean before

**Comment by 7chenko** [comment]
I even changed some text in the code to make sure the changes were getting picked up. No dice, still not scaling up.

**Comment by innovia** [comment]
@7chenko
1) are you sure you pulled from master and have my if added 
2) make sure you run make clean
3) run `go generate ./cloudprovider/aws`
4) do you see if the gpu at 1 on the ec2 instances file?
5) what is your output saying

**Comment by 7chenko** [comment]
Oh I missed the 'go generate' part! Thanks.

**Comment by 7chenko** [comment]
Workaround works for me, thanks! Still waiting on AWS support issue to fix permanently.

**Comment by 7chenko** [comment]
Confirmed that AWS has now fixed the gpu data for p2.xxx:

https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-west-2/index.json

**Comment by innovia** [comment]
@7chenko thanks!

**Comment by bhack** [comment]
@mwielgus Are 0.6.3 or 0.7 released?

**Comment by mwielgus** [comment]
We decided to rename 0.7 as 1.0. It was released, we are also approaching 1.1.

**Comment by bhack** [comment]
@mwielgus Can we deploy 1.0.x with K8 1.7.x?
